apps/hash-ai-worker-py/app/typesystem/entity_type.py

```diff
@@ -1,21 +1,66 @@
 """An entity type schema as defined by the Block Protocol."""
 
-from typing import Any, Literal
+from typing import (
+    Annotated,
+    Any,
+    ClassVar,
+    Literal,
+)
+from uuid import UUID
 
-from pydantic import BaseModel, ConfigDict, Field
+from pydantic import (
+    ConfigDict,
+    Field,
+    GetJsonSchemaHandler,
+    RootModel,
+    create_model,
+)
+from pydantic.json_schema import JsonSchemaValue
+from pydantic_core import CoreSchema
+from slugify import slugify
+from temporalio import workflow
 
-from ._schema import Array, Object, OneOf, OntologyTypeSchema
+from ._schema import Array, Object, OneOf, OntologyTypeSchema, Schema
 from .property_type import PropertyTypeReference
 
 __all__ = ["EntityTypeSchema"]
 
 
-class EntityTypeReference(BaseModel):
+class EntityTypeReference(Schema):
     ref: str = Field(..., alias="$ref")
+    cache: ClassVar[dict[str, type[RootModel]]] = {}
 
+    async def create_model(self, *, actor_id: UUID) -> type[RootModel]:
+        if cached := self.cache.get(self.ref):
+            return cached
 
-class EmptyDict(BaseModel):
-    model_config = ConfigDict(extra="forbid")
+        schema = EntityTypeSchema(
+            **(
+                await workflow.execute_child_workflow(
+                    task_queue="ai",
+                    workflow="getEntityType",
+                    arg={
+                        "entityTypeId": self.ref,
+                        "actorId": actor_id,
+                    },
+                )
+            )["schema"],
+        )
+
+        model = create_model(
+            slugify(self.ref, regex_pattern=r"[^a-z0-9_]+", separator="_"),
+            __base__=RootModel,
+            root=(await schema.create_entity_type(actor_id=actor_id), ...),
+        )
+        self.cache[self.ref] = model
+        return model
+
+
+class EmptyDict(Schema):
+    model_config = ConfigDict(title=None, extra="forbid")
+
+    async def create_model(self, *, actor_id: UUID) -> type[None]:
```

Comment:
shouldn't this return [`Never`](https://docs.python.org/3/library/typing.html#typing.Never)?

Reply:
[`aa90f2a` (#2708)](https://github.com/hashintel/hash/pull/2708/commits/aa90f2a2b59e5443ceef891bc8a272bb9277673e)